File: app/cases.js
/* 案件の画面(app/cases.html)。/api/cases を読んで配信ごとに並べ、状態・メモを /api/cases/update で保存する。
   CSP(script-src 'self')の下で動く。値は textContent で入れる(配信のタイトル・パスに < などが入っていても画面を壊さない) */
(function () {
  'use strict';
  var $ = function (s, r) { return (r || document).querySelector(s); };
  var TOKEN = (document.querySelector('meta[name="ytt-token"]') || {}).content || '';
  var STATUS = { '': '未設定', working: '作業中', posted: '投稿済み', skipped: '見送り' };
  var data = null, toastTimer = null;

  function toast(msg, kind) {
    var t = $('#toast');
    t.className = 'toast ' + (kind || 'info');
    t.textContent = msg;
    t.hidden = false;
    clearTimeout(toastTimer);
    toastTimer = setTimeout(function () { t.hidden = true; }, 4000);
  }
  function err(msg) { var b = $('#errbar'); b.textContent = msg; b.hidden = !msg; }

  function api(path, body) {
    var opt = body ? { method: 'POST', headers: { 'Content-Type': 'application/json', 'X-YTT-Token': TOKEN }, body: JSON.stringify(body) } : {};
    return fetch(path, opt).then(function (r) {
      return r.json().catch(function () { return {}; }).then(function (j) {
        if (!r.ok) throw new Error(j.message || ('HTTP ' + r.status));
        return j;
      });
    });
  }

  function el(tag, cls, text) {
    var e = document.createElement(tag);
    if (cls) e.className = cls;
    if (text != null) e.textContent = text;
    return e;
  }
  function tc(sec) {
    sec = Math.max(0, Math.floor(Number(sec) || 0));
    var h = Math.floor(sec / 3600), m = Math.floor(sec / 60) % 60, s = sec % 60;
    return (h ? h + ':' + String(m).padStart(2, '0') : m) + ':' + String(s).padStart(2, '0');
  }
  function when(ms) {
    if (!ms) return '';
    var d = new Date(ms);
    return (d.getMonth() + 1) + '/' + d.getDate() + ' ' + d.getHours() + ':' + String(d.getMinutes()).padStart(2, '0');
  }
  function link(text, href) {
    var a = el('a', 'btn small', text);
    a.href = href; a.target = '_blank'; a.rel = 'noopener';
    return a;
  }

  function clipRow(c) {
    var li = el('li', 'pt-clip');
    var head = el('div', 'row');
    head.appendChild(el('span', 'pt-clip-range mono', tc(c.start) + '–' + tc(c.end)));
    head.appendChild(el('span', 'grow', c.label || c.file || '(名前なし)'));
    if (!c.exists) head.appendChild(el('span', 'pill warn', '動画が見つからない'));
    li.appendChild(head);
    var st = el('div', 'row pt-clip-steps');
    // 文字起こし
    if (c.transcript) {
      var t = c.transcript, done = t.segments && t.proofed >= t.segments;
      st.appendChild(el('span', 'pill ' + (done ? 'ok' : 'run'), '文字起こし 校正 ' + t.proofed + '/' + t.segments + '行'));
    } else {
      st.appendChild(el('span', 'pill wait', '文字起こし まだ'));
      if (c.exists && c.path) st.appendChild(link('文字起こしで開く', '/transcribe/?media=' + encodeURIComponent(c.path)));
    }
    // パック
    if (c.pack) st.appendChild(el('span', 'pill ok', (c.pack.textplus ? 'Text+ パック ' : 'パック ') + when(c.pack.updatedAt)));
    else {
      st.appendChild(el('span', 'pill wait', 'パック まだ'));
      if (c.exists && c.path) st.appendChild(link('cut2resolve で開く', '/cut2resolve/?video=' + encodeURIComponent(c.path)));
    }
    li.appendChild(st);
    if (c.path) li.appendChild(el('div', 'hint mono pt-clip-path', c.path));
    return li;
  }

  function caseCard(c) {
    var node = $('#tplCase').content.firstElementChild.cloneNode(true);
    node.dataset.id = c.id;
    $('.pt-case-title', node).textContent = c.title || c.id;
    var sub = [c.channel, c.kind === 'youtube' ? 'YouTube' : (c.kind === 'file' ? 'ファイル' : ''),
               'マーク ' + c.marks.total + '(採用 ' + c.marks.adopted + '・書き出し ' + c.marks.exported + ')'];
    if (c.gone) sub.push('スタジオから消えた動画(最後に見えた内容)');
    $('.pt-case-sub', node).textContent = sub.filter(Boolean).join(' ・ ');
    var sel = $('.pt-case-status', node);
    sel.value = c.status || '';
    sel.addEventListener('change', function () {
      api('/api/cases/update', { id: c.id, status: sel.value }).then(function (r) {
        c.status = r.status; toast('状態を「' + STATUS[r.status] + '」にしました', 'ok'); summary();
      }).catch(function (e) { sel.value = c.status || ''; toast('保存できませんでした: ' + e.message, 'err'); });
    });
    var ol = $('.pt-clips', node);
    if (!c.clips.length) ol.appendChild(el('li', 'hint', c.marks.adopted ? '採用したマークはまだ書き出していません(スタジオで書き出すとここに出ます)' : '書き出した切り抜きはありません'));
    c.clips.forEach(function (cl) { ol.appendChild(clipRow(cl)); });
    var ta = $('textarea', node), msg = $('.pt-memo-msg', node);
    ta.value = c.memo || '';
    if (c.memo) $('.pt-case-memo', node).open = true;
    $('.pt-memo-save', node).addEventListener('click', function () {
      api('/api/cases/update', { id: c.id, memo: ta.value }).then(function (r) {
        c.memo = r.memo; msg.textContent = '保存しました';
      }).catch(function (e) { msg.textContent = '保存できませんでした: ' + e.message; });
    });
    return node;
  }

  function visible(c) {
    var fs = $('#fStatus').value, q = $('#fText').value.trim().toLowerCase();
    if (fs === 'none' && c.status) return false;
    if (fs && fs !== 'none' && c.status !== fs) return false;
    if (q && ((c.title || '') + ' ' + (c.channel || '')).toLowerCase().indexOf(q) < 0) return false;
    return true;
  }
  function summary() {
    if (!data) return;
    var n = { '': 0, working: 0, posted: 0, skipped: 0 };
    data.cases.forEach(function (c) { n[c.status || ''] = (n[c.status || ''] || 0) + 1; });
    $('#summary').textContent = '配信 ' + data.cases.length + ' 本(作業中 ' + n.working + '・投稿済み ' + n.posted + '・見送り ' + n.skipped + '・未設定 ' + n[''] + ')';
  }
  function render() {
    var list = $('#list');
    list.textContent = '';
    var shown = data.cases.filter(visible);
    if (!shown.length) list.appendChild(el('p', 'empty', data.cases.length ? '条件に合う配信はありません' : 'まだ配信がありません(切り抜きスタジオで配信を解析すると、ここに出ます)'));
    shown.forEach(function (c) { list.appendChild(caseCard(c)); });
    var ul = $('#unlinked');
    ul.textContent = '';
    data.unlinked.forEach(function (t) {
      var li = el('li', '');
      li.appendChild(el('span', '', (t.title || t.id) + '(校正 ' + t.proofed + '/' + t.segments + '行)'));
      if (t.sourcePath) li.appendChild(el('div', 'hint mono', t.sourcePath));
      ul.appendChild(li);
    });
    $('#unlinkedCount').textContent = data.unlinked.length + ' 件';
    $('#unlinkedBox').hidden = !data.unlinked.length;
    $('#casesFile').textContent = '状態・メモの保存先: ' + (data.casesFile || '');
    summary();
  }
  function load() {
    err('');
    return api('/api/cases').then(function (j) { data = j; render(); })
      .catch(function (e) { err('案件の一覧を読めませんでした: ' + e.message); });
  }

  $('#fStatus').addEventListener('change', render);
  $('#fText').addEventListener('input', function () { if (data) render(); });
  $('#btnReload').addEventListener('click', function () { load().then(function () { toast('読み込み直しました'); }); });
  load();
})();

File: app/cases.py
"""案件(配信1本)ごとの紐づけ(統合計画の段階4)。入口の「案件」画面(app/cases.html)の中身を作る。

案件 = 切り抜きスタジオの動画1本(配信・ファイル)。その下に、書き出した切り抜きと、それぞれの文字起こし・パックを並べる。
紐づけは、各ツールが持っているデータを**読むたびに組み立て直す**(各ツールの記録と食い違わないように。2026-09-26 ユーザーと決定):
  - 切り抜き … スタジオの data.json の「書き出し済み」のマーク(書き出した mp4 の絶対パスを持つ)
  - 文字起こし … 文字起こしツールの transcripts/*.json(元の動画のパスが切り抜きと同じ。無ければ .clip.json の配信・マークが同じ。
                規則は ytt_core/txindex.py にあり、スタジオのセリフの表示と共通)
  - パック … 切り抜きの隣の <名前>_pack フォルダ(cut2resolve の既定の出力先)
案件ファイル(<作業データ>\\app\\cases.json)に持つのは、人が付ける状態・メモと、最後に見えた紐づけ(元のファイルを消しても履歴が残るように)。
各ツールのデータは読むだけで、書き換えない。
"""
import json
import os
import threading
import time

from ytt_core import datadir, fsio, txindex

SCHEMA = "youtube-tools-cases/v1"
STATUSES = ("", "working", "posted", "skipped")        # 未設定・作業中・投稿済み・見送り
STATUS_LABELS = {"": "未設定", "working": "作業中", "posted": "投稿済み", "skipped": "見送り"}
MAX_MEMO = 2000
MAX_JSON = 64 * 1024 * 1024
_lock = threading.Lock()


def _read_json(path, limit=MAX_JSON):
    try:
        return fsio.read_json_file(path, limit)
    except (OSError, UnicodeError, ValueError):
        return None


def locations(repo_root, env=None):
    """{"studio": data.json, "transcripts": フォルダ, "cases": cases.json}。各ツールと同じ規則で置き場所を決める"""
    env = os.environ if env is None else env
    studio_home = env.get("STUDIO_HOME") or datadir.tool_dir("studio", os.path.join(repo_root, "clip-studio"), env)
    app_home = datadir.tool_dir("app", os.path.join(repo_root, "app"), env)
    return {"studio": os.path.join(studio_home, "data.json"), "transcripts": txindex.folder(repo_root, env),
            "cases": os.path.join(app_home, "cases.json")}


# ---------------------------------------------------------------- 各ツールのデータを読む(読むだけ)

def read_studio(path):
    d = _read_json(path)
    videos = d.get("videos") if isinstance(d, dict) else None
    return videos if isinstance(videos, dict) else {}


def read_transcripts(folder):
    """文字起こしの一覧(ytt_core.txindex。スタジオのセリフの表示と同じ読み方・紐づけの規則)"""
    return txindex.load(folder)


def find_pack(media_path):
    """切り抜きの隣の <名前>_pack(cut2resolve の既定の出力先)。-> {"dir", "textplus", "updatedAt"} か None"""
    if not media_path:
        return None
    d = os.path.join(os.path.dirname(media_path), os.path.splitext(os.path.basename(media_path))[0] + "_pack")
    plan = os.path.join(d, "cut-plan.json")
    if not os.path.isfile(plan):
        return None
    try:
        mt = int(os.path.getmtime(plan) * 1000)
    except OSError:
        mt = 0
    return {"dir": d, "textplus": os.path.isfile(os.path.join(d, "textplus-import.json")), "updatedAt": mt}


# ---------------------------------------------------------------- 組み立て

def build(videos, transcripts, saved=None, pack_finder=find_pack):
    """-> {"cases": [...], "unlinked": [文字起こし]}。saved: cases.json の中身(状態・メモ・最後に見えた紐づけ)"""
    saved = saved or {}
    used = set()
    cases = []
    for vid, v in videos.items():
        if not isinstance(v, dict):
            continue
        marks = [m for m in (v.get("marks") or []) if isinstance(m, dict)]
        clips = []
        for m in sorted((m for m in marks if m.get("status") == "exported"), key=lambda m: (m.get("start") or 0)):
            path = m.get("path") if isinstance(m.get("path"), str) else ""
            tx, n, ids = txindex.pick(transcripts, vid, m.get("id"), path)
            used.update(ids)
            clips.append({"markId": str(m.get("id") or ""), "label": str(m.get("label") or "")[:80], "start": m.get("start"), "end": m.get("end"),
                          "file": str(m.get("file") or ""), "path": path, "exists": bool(path) and os.path.isfile(path),
                          "transcript": txindex.summary(tx) if tx else None, "transcripts": n, "pack": pack_finder(path) if path else None})
        s = saved.get(vid) if isinstance(saved.get(vid), dict) else {}
        cases.append({"id": vid, "kind": v.get("kind") or "", "title": str(v.get("title") or v.get("fileName") or vid)[:120],
                      "channel": str(v.get("channel") or "")[:100], "duration": v.get("duration") or 0,
                      "marks": {"total": len(marks), "adopted": sum(1 for m in marks if m.get("status") == "adopted"),
                                "exported": len(clips), "candidates": sum(1 for m in marks if not m.get("status"))},
                      "clips": clips, "status": s.get("status") if s.get("status") in STATUSES else "",
                      "memo": str(s.get("memo") or "")[:MAX_MEMO], "statusUpdatedAt": s.get("statusUpdatedAt") or 0,
                      "updatedAt": v.get("updatedAt") or 0, "gone": False})
    # スタジオから消えた動画も、状態・メモを付けていれば、最後に見えた紐づけで残す
    for cid, s in saved.items():
        if cid not in videos and isinstance(s, dict) and (s.get("status") or s.get("memo")) and isinstance(s.get("last"), dict):
            last = dict(s["last"], id=cid, status=s.get("status") if s.get("status") in STATUSES else "",
                        memo=str(s.get("memo") or "")[:MAX_MEMO], gone=True)
            cases.append(last)
    cases.sort(key=lambda c: -(c.get("updatedAt") or 0))
    unlinked = [dict(txindex.summary(t), sourcePath=t["sourcePath"]) for t in transcripts if t["id"] not in used]
    unlinked.sort(key=lambda t: -t["updatedAt"])
    return {"cases": cases, "unlinked": unlinked}


# ---------------------------------------------------------------- 案件ファイル

def load_saved(path):
    d = _read_json(path, 16 * 1024 * 1024)
    if not isinstance(d, dict) or d.get("schema") != SCHEMA or not isinstance(d.get("cases"), dict):
        return {}
    return d["cases"]


def _write(path, saved):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    fsio.write_json(path, {"schema": SCHEMA, "cases": saved})


def snapshot(repo_root, env=None):
    """画面用の一覧。最後に見えた紐づけ(状態・メモを付けた案件だけ)を案件ファイルに残す"""
    loc = locations(repo_root, env)
    with _lock:
        saved = load_saved(loc["cases"])
        res = build(read_studio(loc["studio"]), read_transcripts(loc["transcripts"]), saved)
        changed = False
        for c in res["cases"]:
            s = saved.get(c["id"])
            if s is not None and not c["gone"]:
                last = {k: c[k] for k in ("kind", "title", "channel", "duration", "marks", "clips", "updatedAt")}
                if s.get("last") != last:
                    s["last"] = last
                    changed = True
        if changed:
            try:
                _write(loc["cases"], saved)
            except OSError:
                pass
    res["casesFile"] = loc["cases"]
    return res


def update(repo_root, case_id, status=None, memo=None, env=None):
    """状態・メモを付ける。-> 更新後の {status, memo, statusUpdatedAt}"""
    if not isinstance(case_id, str) or not (1 <= len(case_id) <= 64) or not all(ch.isalnum() or ch in "-_" for ch in case_id):
        raise ValueError("案件の指定が正しくありません")
    if status is not None and status not in STATUSES:
        raise ValueError("状態が正しくありません")
    if memo is not None and (not isinstance(memo, str) or len(memo) > MAX_MEMO):
        raise ValueError("メモは %d 文字までです" % MAX_MEMO)
    loc = locations(repo_root, env)
    with _lock:
        saved = load_saved(loc["cases"])
        s = saved.setdefault(case_id, {})
        if status is not None:
            s["status"] = status
            s["statusUpdatedAt"] = int(time.time() * 1000)
        if memo is not None:
            s["memo"] = memo
        if not s.get("status") and not s.get("memo"):
            saved.pop(case_id, None)   # 何も付けていない案件は持たない(ファイルを小さく)
        _write(loc["cases"], saved)
        return {"status": s.get("status", ""), "memo": s.get("memo", ""), "statusUpdatedAt": s.get("statusUpdatedAt", 0)}
